plugins/inputs/pulsar_consumer/pulsar_consumer.go

```diff
@@ -0,0 +1,248 @@
+  ## output metric_batch_size is 1000, setting this to 100 will ensure that a
+  ## full batch is collected and the write is triggered immediately without
+  ## waiting until the next flush_interval.
+  # max_undelivered_messages = 1000
+
+  ## Data format to consume.
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+  data_format = "influx"
+`
+
+func (p *pulsarConsumer) SampleConfig() string {
+	return sampleConfig
+}
+
+func (p *pulsarConsumer) Description() string {
+	return "Read metrics from Pulsar topic"
+}
+
+func (p *pulsarConsumer) SetParser(parser parsers.Parser) {
+	p.parser = parser
+}
+
+func (p *pulsarConsumer) Start(acc telegraf.Accumulator) error {
+	p.acc = acc.WithTracking(p.MaxUndeliveredMessages)
+	ctx, cancel := context.WithCancel(context.Background())
+	p.cancel = cancel
+
+	if p.connPool == nil && p.consumer == nil {
+		p.in = make(chan pulsar.Message, p.MaxUndeliveredMessages)
+		p.errs = make(chan error)
+
+		var err error
+		p.recvTimeout, err = time.ParseDuration(p.RecvTimeout)
+
+		conf := pulsar.ManagedConsumerConfig{}
+		conf.Addr = p.URL
+		conf.Errs = p.errs
+		conf.DialTimeout, err = time.ParseDuration(p.DialTimeout)
+		conf.ConnectTimeout = conf.DialTimeout
+		conf.PingFrequency, err = time.ParseDuration(p.PingFrequency)
+		conf.PingTimeout, err = time.ParseDuration(p.PingTimeout)
+		conf.InitialReconnectDelay, err = time.ParseDuration(p.InitialReconnectDelay)
+		conf.MaxReconnectDelay, err = time.ParseDuration(p.MaxReconnectDelay)
+		conf.NewConsumerTimeout, err = time.ParseDuration(p.NewConsumerTimeout)
+		if p.TLSCA != "" && p.TLSCert != "" && p.TLSKey != "" {
+			conf.TLSConfig, err = p.TLSConfig()
+		}
+		if err != nil {
+			return err
+		}
+
+		conf.Topic = p.Topic
+		conf.Name = p.Name
+		conf.Exclusive = p.Exclusive
+		conf.QueueSize = p.QueueSize
+
+		p.connPool = pulsar.NewManagedClientPool()
+		p.consumer = pulsar.NewManagedConsumer(p.connPool, conf)
+
+		go func() {
+			err = p.consumer.ReceiveAsync(ctx, p.in)
+			if err != nil {
+				p.errs <- err
+			}
+		}()
+	}
+
+	// Start the message reader
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		go p.receiver(ctx)
+	}()
+
+	log.Printf("I! Started the Pulsar consumer service, pulsar: %v, topic: %v, name: %v\n",
+		p.URL, p.Topic, p.Name)
+
+	return nil
+}
+
+func (p *pulsarConsumer) receiver(ctx context.Context) {
+	sem := make(semaphore, p.MaxUndeliveredMessages)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-p.acc.Delivered():
+			<-sem
+		case err := <-p.errs:
+			p.acc.AddError(err)
+		case sem <- empty{}:
+			select {
+			case <-ctx.Done():
+				return
+			case err := <-p.errs:
+				<-sem
+				log.Printf("E! Unexpected error, pulsar: %v error: %v, topic: %v, name: %v\n",
+					p.URL, err, p.Topic, p.Name)
+				p.acc.AddError(err)
```

Comment:
This will end up double logging, remove `AddError` or send it the above printed error